File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.cpp
class Solution {
public:
    void bcktrack(int n, int len, vector<char>& tmp, vector<string>* ans){
        if(n == len) {
            ans->push_back(string{tmp.begin(), tmp.end()});
            return;
        }
        for(char i = 'a'; i <= 'c'; ++i){
            if(!tmp.empty() && *tmp.rbegin() == i) continue;
            tmp.push_back(i);
            bcktrack(n, len + 1, tmp, ans);
            tmp.pop_back();
        }
    }

    string getHappyString(int n, int k) {
        vector<string> res{};
        vector<char> tmp;
        bcktrack(n, 0, tmp, &res);

        return res.size() < k ? "" : res[k - 1];
    }
};

File: 1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n/1516-the-k-th-lexicographical-string-of-all-happy-strings-of-length-n.erl
-spec get_happy_string(N :: integer(), K :: integer()) -> unicode:unicode_binary().
get_happy_string(N, K) ->
    list_to_binary(element(2, bcktrack([$a, $b, $c], N, K, 0, 1, -1, []))).

bcktrack(_, N, K, N, K, _, Res) -> {K + 1, Res};
bcktrack(_, N, K, N, Th, _, Res) -> 
    {Th + if Th < K -> 1; true -> 0 end, []};
bcktrack(List, N, K, Len, Th, Prev, Res) ->
    lists:foldl(
        fun(Ch, {_Th, _Res}) ->
            if
                (Ch == Prev) or (_Th > K) -> {_Th, _Res};
                true -> 
                    % {__Th, __Res} = bcktrack(List, N, K, Len + 1, _Th, Ch, Res ++ [Ch]),
                    % {__Th, case __Res of
                    %     [] -> _Res;
                    %     _ -> __Res
                    % end}
                    bcktrack(List, N, K, Len + 1, _Th, Ch, Res ++ [Ch])
            end
        end, {Th, []}, List
    ).